pertpy/tools/_distances/_distances.py

```diff
@@ -29,6 +30,113 @@
     from anndata import AnnData
 
 
+@jit(nopython=True, parallel=True, cache=True)
+def _euclidean_pairwise_mean_within(X: np.ndarray) -> float:
+    """Compute mean pairwise euclidean distance within a group (X to X).
+
+    Efficiently computes the mean without storing the full distance matrix.
+    Uses parallel computation for speed.
+
+    Args:
+        X: Array of shape (n_samples, n_features).
+
+    Returns:
+        Mean pairwise euclidean distance within the group.
+    """
+    n_samples = X.shape[0]
+    if n_samples < 2:
+        return 0.0
+
+    total_distance = 0.0
+    n_pairs = n_samples * (n_samples - 1) / 2.0
+
+    # Compute all pairs (i, j) where i < j to avoid duplicates and self-distances
+    for i in prange(n_samples):
+        for j in range(i + 1, n_samples):
+            # Compute euclidean distance
+            dist_sq = 0.0
+            for k in range(X.shape[1]):
+                diff = X[i, k] - X[j, k]
+                dist_sq += diff * diff
+            total_distance += np.sqrt(dist_sq)
+
+    return total_distance / n_pairs
+
+
+@jit(nopython=True, parallel=True, cache=True)
+def _euclidean_pairwise_mean_between(X: np.ndarray, Y: np.ndarray) -> float:
+    """Compute mean pairwise euclidean distance between two groups (X to Y).
+
+    Efficiently computes the mean without storing the full distance matrix.
+    Uses parallel computation for speed.
+
+    Args:
+        X: Array of shape (n_samples_X, n_features).
+        Y: Array of shape (n_samples_Y, n_features).
+
+    Returns:
+        Mean pairwise euclidean distance between the groups.
+    """
+    n_samples_X = X.shape[0]
+    n_samples_Y = Y.shape[0]
+
+    if n_samples_X == 0 or n_samples_Y == 0:
+        return 0.0
+
+    total_distance = 0.0
+    n_pairs = n_samples_X * n_samples_Y
+
+    # Compute all pairs between X and Y
+    for i in prange(n_samples_X):
+        for j in range(n_samples_Y):
+            # Compute euclidean distance
+            dist_sq = 0.0
```

Comment:
This part here is duplicated with the one above, right? I'm okay with it because it's explicit here.